tests/cxx/virtual_methods/virtual_call_in_ctor_ub.cpp

```diff
@@ -0,0 +1,91 @@
+struct AVirtual
+{
+  int m_a;
+
+  AVirtual(AVirtual *a) : m_a(0)
+  //@ requires true;
+  //@ ensures AVirtual_vtype(this, &typeid(struct AVirtual)) &*& this->m_a |-> 0;
+  {
+    foo();
+  }
+
+  virtual ~AVirtual()
+  //@ requires AVirtual_vtype(this, &typeid(struct AVirtual)) &*& this->m_a |-> _;
+  //@ ensures true;
+  {
+    foo();
+  }
+
+  int getI() const
+  //@ requires this->m_a |-> ?a;
+  //@ ensures this->m_a |-> a &*& result == a;
+  {
+    return m_a;
+  }
+
+  virtual void foo() const
+  //@ requires AVirtual_vtype(this, ?t);
+  //@ ensures AVirtual_vtype(this, t);
+  {}
+};
+
+struct BVirtual
+{
+  int m_b;
+
+  BVirtual(AVirtual *av) : m_b(av->m_a)
+  //@ requires av != 0 &*& av->m_a |-> ?a;
+  //@ ensures av->m_a |-> a &*& BVirtual_vtype(this, &typeid(struct BVirtual)) &*& this->m_b |-> a;
+  {
+    av->foo(); //~
```

Comment:
When verifying a virtual method implementation for struct `S`, a `S_vtype` chunk is consumed (like we do for `S_bases_constructed` when `S` derives from at least one struct) and reproduced before the body is executed. Hence, `foo` requires that chunk in its precondition.

Reply:
I see. I had expected that the existence of the `S_vtype` chunk would be checked at the call site (in analogy with what happens at run time: the vtable pointer is dereferenced at the call site to get the pointer to the function to call), but I guess that indeed, at verification time, the call site does not care about the value of the vtable pointer since it is always the statically resolved spec that is used to verify it, so the vtable pointer existence check might as well happen in the callee. OK; looks good!